Library/Homebrew/github_packages.rb

```diff
@@ -38,35 +39,131 @@ def upload_bottles(bottles_hash)
+    ])
+
+    oras = HOMEBREW_PREFIX/"bin/oras"
+    unless oras.exist?
       ohai "Installing `oras` for upload..."
       safe_system HOMEBREW_BREW_FILE, "install", "oras"
-      Formula["oras"].opt_bin/"oras"
+      oras = Formula["oras"].opt_bin/"oras"
     end
 
     bottles_hash.each do |formula_name, bottle_hash|
       _, org, repo, = *bottle_hash["bottle"]["root_url"].match(URL_REGEX)
+
+      # docker CLI insists on lowercase org ("repository name")
+      org = org.downcase
+      image = "#{URL_DOMAIN}/#{org}/#{repo}/#{formula_name}"
+
       version = bottle_hash["formula"]["pkg_version"]
-      rebuild = bottle_hash["bottle"]["rebuild"]
+      rebuild = if (rebuild = bottle_hash["bottle"]["rebuild"]).positive?
+        ".#{rebuild}"
+      end
 
-      bottle_hash["bottle"]["tags"].each do |bottle_tag, tag_hash|
+      formula_path = HOMEBREW_REPOSITORY/bottle_hash["formula"]["path"]
+      formula = Formulary.factory(formula_path)
+
+      image_tags = bottle_hash["bottle"]["tags"].map do |bottle_tag, tag_hash|
         local_file = tag_hash["local_filename"]
         odebug "Uploading #{local_file}"
 
-        tag = "#{version}.#{bottle_tag}"
-        tag = "#{tag}.#{rebuild}" if rebuild.positive?
-
+        tag = "#{version}.#{bottle_tag}#{rebuild}"
+
+        # TODO: ideally most/all of these attributes would be stored in the
+        # bottle JSON rather than reading them from the formula.
+        git_revision = formula.tap.git_head
+        git_path = formula_path.to_s.delete_prefix("#{formula.tap.path}/")
+        manifest_hash = {
+          "org.opencontainers.image.title"    => formula.full_name,
+          "org.opencontainers.image.url"      => formula.homepage,
+          "org.opencontainers.image.version"  => version,
+          "org.opencontainers.image.revision" => git_revision,
+          "org.opencontainers.image.source"   => "https://github.com/#{org}/#{repo}/blob/#{git_revision}/#{git_path}",
+        }
+        manifest_hash["org.opencontainers.image.description"] = formula.desc if formula.desc.present?
+        manifest_hash["org.opencontainers.image.license"] = formula.license if formula.license.present?
+
+        manifest_annotations = Pathname("#{formula_name}.#{tag}.annotations.json")
+        manifest_annotations.unlink if manifest_annotations.exist?
+        manifest_annotations.write({ "$manifest" => manifest_hash }.to_json)
+
+        # TODO: ideally most/all of these attributes would be stored in the
+        # bottle JSON rather than reading them from the formula.
+        os, arch, os_version = if OS.mac?
+          os = "darwin"
+          macos_version = MacOS::Version.from_symbol(bottle_tag.to_sym)
+          arch = if macos_version.arch == :arm64
+            "arm64"
+          else
+            "amd64"
+          end
+          [os, arch, macos_version.to_f.to_s]
+        elsif OS.linux?
+          ["linux", "amd64", OS.kernel_version.to_s]
```

Comment:
```suggestion
          ["linux", "amd64", SystemConfig.host_glibc_version.to_s]
```
The version of Glibc is more relevant to bottle compatibility than the Linux kernel version.

It should be the version of `glibc` used to build the bottle, not the version on the machine uploading the bottle. The version of `glibc` used to build the bottle is not stored in the install receipt currently, though it would be nice to have that info. The `"built_on": "os_version": "Ubuntu 16.04.7 LTS"` is stored in the bottle.

Reply:
> The version of Glibc is more relevant to bottle compatibility than the Linux kernel version.

It's not the OS version, though, or the one that's being used in these images. I'll stick to using the `os_version` from the bottle `INSTALL_RECEIPT`.